src/resolve-names.cc

```diff
@@ -412,7 +412,21 @@ Result NameResolver::OnCatchExpr(TryExpr*, Catch* catch_) {
 }
 
 Result NameResolver::OnDelegateExpr(TryExpr* expr) {
-  ResolveLabelVar(&expr->delegate_target);
+  Var* delegate_target = &expr->delegate_target;
+  bool was_name = delegate_target->is_name();
+  ResolveLabelVar(delegate_target);
+  if (was_name) {
+    // Invalid label (of the immediate try-delegate block) used.
+    if (delegate_target->index() == 0) {
+      PrintError(&delegate_target->loc, "invalid label variable for try-delegate");
+      return Result::Ok;
+    }
+    // Adjust label by one, because a delegate instruction cannot target its
+    // immediately enclosing try block.
+    delegate_target->set_index(delegate_target->index() - 1);
+  }
+  // Pop the label here as a try-delegate has no `end` instruction.
+  PopLabel();
```

Comment:
I might be mistaken, but if we do this `PopLabel()` at the top, wouldn't we not need all these fixup routines?

The point of the current rule is to think of the whole `try`-`delegate` like a single branch instruction, and treat its immediate argument as such.

Reply:
Good point, I think that will work. I pushed a commit that makes this change. It also changes the error message on the failing test case a bit, but I think that should be fine.